webapi/src/main/scala/org/knora/webapi/util/search/gravsearch/GravsearchUtilV2.scala

```diff
@@ -70,6 +403,97 @@ object GravsearchUtilV2 {
+                    StatementPattern.makeExplicit(subj = resourceValueObject, pred = resourceValueObjectProp, obj = resourceValueObjectObj)
+                )
+
+                // return assertions about value objects
+                val constructPatternsForValueObjects = Seq(
+                    StatementPattern(subj = resourceVar, pred = IriRef(OntologyConstants.KnoraBase.HasValue.toSmartIri), obj = resourceValueObject),
+                    StatementPattern(subj = resourceVar, pred = resourceValueProp, obj = resourceValueObject),
+                    StatementPattern(subj = resourceValueObject, pred = resourceValueObjectProp, obj = resourceValueObjectObj)
+                )
+
+                // WHERE patterns for standoff belonging to value objects (if any)
+                val wherePatternsForStandoff = Seq(
+                    ValuesPattern(resourceValueObject, valueObjectIris.map(iri => IriRef(iri.toSmartIri))),
+                    StatementPattern.makeExplicit(subj = resourceValueObject, pred = IriRef(OntologyConstants.KnoraBase.ValueHasStandoff.toSmartIri), obj = standoffNodeVar),
+                    StatementPattern.makeExplicit(subj = standoffNodeVar, pred = standoffPropVar, obj = standoffValueVar)
+                )
+
+                // return standoff
+                val constructPatternsForStandoff = Seq(
+                    StatementPattern(subj = resourceValueObject, pred = IriRef(OntologyConstants.KnoraBase.ValueHasStandoff.toSmartIri), obj = standoffNodeVar),
+                    StatementPattern(subj = standoffNodeVar, pred = standoffPropVar, obj = standoffValueVar)
+                )
+
+                ConstructQuery(
+                    constructClause = ConstructClause(
+                        statements = constructPatternsForResources ++ constructPatternsForValueObjects ++ constructPatternsForStandoff
+                    ),
+                    whereClause = WhereClause(
+                        Seq(
+                            UnionPattern(
+                                Seq(wherePatternsForResources, wherePatternsForValueObjects, wherePatternsForStandoff)
+                            )
+                        )
+                    )
+                )
+
+            } else {
+                // no value objects are to be queried
+
+                ConstructQuery(
+                    constructClause = ConstructClause(
+                        statements = constructPatternsForResources
+                    ),
+                    whereClause = WhereClause(
+                        Seq(
+                            UnionPattern(
+                                Seq(wherePatternsForResources)
+                            )
+                        )
+                    )
+                )
+            }
+
+        }
+
     }
 
     object Gravsearch {
```

Comment:
What's the difference between `object Gravsearch` and `object GravsearchUtilV2`? If they're not really distinct, please merge them. Otherwise, please put them in separate files.

Reply:
I tried to distinguish between fulltext and Gravsearch mode. `GravsearchUtilV2` contains them both. A fulltext search is not really a Gravsearch from the user's perspective, but some of the logic used is similar.

Reply:
Instead of calling it `GravsearchUtilV2` or `SearchResponderUtilV2`, which are both pretty vague, could we break this file up into several parts with specific names? Are there distinct subsets of functionality here, each of which could be described in a class or object with a Scaladoc comment?

Reply:
I mean, break this file up into several files?

Reply:
There is definitely a part that handles Sparql conversions (mainly used by the transformation classes).

And there are parts used to process query results. I guess we could split them into separate files, yes.

Reply:
That sounds good to me.

Reply:
At the very least I think those two parts (SPARQL conversions and result processing) should be in separate files. And maybe each part also has subparts that could be separated, too.

Reply:
Ok, and the we would write separate specs for them.

Reply:
I think I can do the splitting up next week. Specs will come step by step as we work on bugs and new features.

Reply:
OK great.